### ISSUE TITLE
Example tasks for Visual Studio Code

### ISSUE THREAD

**Issue by jcotela** [KratosMultiphysics/Kratos#579]
Dear all, I have been playing a bit with tasks in Visual Studio Code to have automated compilation and testing of Kratos. @pooyan-dadvand asked me to post my configuration file, thinking that it may be useful for other developers.

Some notes:
- Run a task with `ctrl + p` followed by typing `task <task-name>` on the resulting box
- Unlike the usual guidelines, I don't have the `make -j4 install`instruction on my kratos configuration script, hence the two tasks (configure only does the cmake stuff, then build calls `make`).
- It may be more convenient to assign [key bindings](https://code.visualstudio.com/docs/editor/tasks#_binding-keyboard-shortcuts-to-tasks) to the tasks for convenience
- All paths correspond to my local installation. Change accordingly!
- I've done this for a relatively recent version of Visual Studio Code (1.13.1), it may not work for older version, since it needs the new task engine (2.0.0)
- the configure script (`configure.sh` in my case) needs execute permissions to be called like this.
- You may need to tell the shell within visual studio code to read your bashrc (mine didn't by default) in order to run your tests, otherwise it won't find the kratos installation. If this happens, add the following entry to your user settings: `"terminal.integrated.shellArgs.linux": ["-i" ]` 

The file itself: `tasks.json`

```
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks" : [
        {
            "taskName": "configure-release",
            "command": "./configure.sh",
            "args": [],
            "isShellCommand": true,
            "echoCommand": false,
            "options": {
                "cwd": "${workspaceRoot}/build_release"
            }
        },
        {
            "taskName": "build-release",
            "command": "make",
            "args": [ "-j4", "install" ],
            "isShellCommand": true,
            "isBuildCommand": true,
            "dependsOn": "release-configure",
            "problemMatcher": {
                "owner": "cpp",
                "fileLocation": "absolute",
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5
                }
            },
            "options": {
                "cwd": "${workspaceRoot}/build_release"
            }
        },
        {
            "taskName": "run-c-tests",
            "command": "python",
            "args": [ "kratos_c_tests.py" ],
            "isShellCommand": true,
            "options": {
                "cwd": "${workspaceRoot}/applications/TurbulentFlowApplication/python_scripts"
            }
        },
        {
            "taskName": "run-python-tests",
            "command": "python",
            "args": [ "run_tests.py", "-c\"python\"" ],
            "isShellCommand": true,
            "options": {
                "cwd": "${workspaceRoot}/kratos/python_scripts"
            }
        }
    ]
}
```

**Comment by pooyan-dadvand** [comment]
Thank you very much! Let see if there is some feedback and improvement and then we can add it to wiki

**Comment by asroy** [comment]
Hi @jcotela @pooyan-dadvand,

I happened to have spent some time trying to use VS code as the front end for python and C++ debuggers, so I can do things like setting breakpoints and stepping through python and C++ portion of Kratos code in a single GUI. I set it up in the way so that VS code would let python debugger launch a python process or attach to it, and then let GDB attach to it. I can do the same things if using mpirun to launch more than one processes. Each debugger for each process would have their own call stacks, but the VS code organizes them nicely. I found it to be very useful in stepping through code to understand it.

I can write down how to set it up if you think it would be useful.

Chao

![screenshot_20170704_190111](https://user-images.githubusercontent.com/22615726/27844726-357e216e-60eb-11e7-9b98-76bd1d37b742.png)

**Comment by jcotela** [comment]
Hi @asroy 
If you could write a small how-to, it would be definitely very helpful.

**Comment by philbucher** [comment]
@asroy this would be very helpful, thanks a lot!

**Comment by asroy** [comment]
@jcotela @philbucher @pooyan-dadvand
Ok, I will write it down this evening. Thanks
Chao

**Comment by maceligueta** [comment]
I have to try it! Sounds awesome. 
Right now I was debugging Python with Spyder and C++ with GDB...

**Comment by asroy** [comment]
@jcotela @pooyan-dadvand @philbucher @maceligueta 

I've tested it on Fedora 23 and Ubuntu 16.04. I'm using gcc-5.3.1, python-3.4, VS Code-1.31.1
You can debug mixed Python/C++ in the same GUI. It also works for MPI applications. You can switch between the debuggers and corresponding call stacks.

1. Packages needed
  1) Visual Studio Code

  2) Extensions for VS Code:
      "Python" from Don Jayamanne (I'm using 0.6.7)
        This allows VS Code act as the front end to debug python.
        This gives VS Code ability to attach to a python script that uses module "ptvsd". 
  
      "C/C++" from Microsoft (I'm using 0.12.0)
        This allows VS Code act as the front end for gdb/lldb
  
  3) python module:
      ptvsd-3.0.0
        When use this module in your python script, VS Code can attach to the script.
        The lastest version (3.1.0) doesn't work with the VS Code extension "Python"
        you can install it using pip
  

2. Setting up VS Code for Python/C++ debugging
  1) Senario 1: For non-MPI application
    a) First, in VS Code, launch a python script with python debugger:
      See Appendix B
      Start debugger mode "Python launch"

    b) Then, in VS code, attach GDB to the python process:
      See Appendix B
      Start debugger mode "GDB Attach proc 0", see Appendix below

  2) Senario 2: For MPI application, attach python debugger and GDB to each process
    a) In the python script your want to run, added the code in Appendix A to use ptvsd
      ptvsd allow a debugger from remote machine to attach to a python script running on the local machine
      We are going to run both the python script and the debugger on local machine. So ip address is ip address of the local machine
      For the port number, if you are using multiple MPI process, assign a unused port to each process

    b) Then, in terminal, use mpiexec to launch the python script
      mpiexec -np 2 xterm -e python3 your_python_script.py
      2 MPI processes will be launched. The processes would wait on "ptvsd.wait_for_attach()", I use port number 3000 for process 0, and 3001 for process 1
    
    c) Then in VS Code, attach python debugger to the python process
      See Appendix B
      Start debugger mode "Python Attach (local) proc 0". This python debugger will attach to the MPI process 0 
      Start debugger mode "Python Attach (local) proc 1". This python debugger will attach to the MPI process 1
    
    d) Then in VS Code, attach GDB to the python process
      See Appendix B
      Start debugger mode "GDB Attach proc 0". VS Code will ask you which process you want to attach to, select the one that is MPI process 0
      Start debugger mode "GDB Attach proc 1". VS Code will ask you which process you want to attach to, select the one that is MPI process 1


You only need to add ptvsd in the top most python script.
If you only want to debug one MPI process, you only need to use ptvsd on that process. 
If you want to debug more MPI processes than 2 in the example, add more debugger configurations in the launch.json file.
Remember to use unique port number for each MPI process you want to debug (in python script and also in launch.json file)


 
Appendix A: python code to enable Python debugger attach
====================== begining added code to enable ptvsd ========================
import ptvsd
port=3000
ptvsd.enable_attach(secret='my_secret', address =('127.0.0.1', port))
ptvsd.wait_for_attach()
======================== end of added code to enable ptvsd =========================
                    


Appendix B: configure file "launch.json" for VS Code
====================== begining of launch.json file ===============================
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python launch",
            "type": "python",
            "request": "launch",
            "stopOnEntry": true,
            "pythonPath": "/usr/bin/python3",
            "program": "${workspaceRoot}/MainKratos.py",
            "cwd": "",
            "console": "externalTerminal",
            "env": {},
            "envFile": "${workspaceRoot}/.env",
            "debugOptions": [
                "WaitOnAbnormalExit",
                "WaitOnNormalExit"
            ]
        },
        {
            "name": "Python Attach (Remote) proc 0",
            "type": "python",
            "request": "attach",
            "localRoot": "${workspaceRoot}",
            "remoteRoot": "${workspaceRoot}",
            "port": 3000,
            "secret": "my_secret",
            "host": "localhost"
        },
        {
            "name": "Python Attach (Remote) proc 1",
            "type": "python",
            "request": "attach",
            "localRoot": "${workspaceRoot}",
            "remoteRoot": "${workspaceRoot}",
            "port": 3001,
            "secret": "my_secret",
            "host": "localhost"
        },
        { 
            "name": "GDB Attach proc 0",
            "type": "cppdbg",
            "request": "attach",
            "program": "/usr/bin/python3",
            "processId": "${command:pickProcess}",
            "MIMode": "gdb"
        },
        { 
            "name": "GDB Attach proc 1",
            "type": "cppdbg",
            "request": "attach",
            "program": "/usr/bin/python3",
            "processId": "${command:pickProcess}",
            "MIMode": "gdb"
        }
    ]
}
======================== end of launch.json file ===================================

**Comment by asroy** [comment]
@jcotela @pooyan-dadvand @philbucher @maceligueta 
I cannot get text format correct here. So I moved it here:
https://gist.github.com/asroy/ca018117e5dbbf53569b696a8c89204f

The example with MPI is attached here:
[cavity_3d.zip](https://github.com/KratosMultiphysics/Kratos/files/1129519/cavity_3d.zip)

There is a .vscode folder that contains the configuration file for VS Code
The ptvsd module that needs to be used to enable python debugger attach, is enabled in MainKratos.py

I also wrote down the process I built Kratos with MPI enabled, I'm attaching it here:
[build_kratos_mpi-instruction.zip](https://github.com/KratosMultiphysics/Kratos/files/1129561/build_kratos_mpi-instruction.zip)


Let me know if the setup is not working for you

**Comment by jcotela** [comment]
Thanks a lot for taking the time to write this @asroy! I'll give it a try later today.

**Comment by RiccardoRossi** [comment]
closing as long inactive
